src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -15,53 +15,105 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+        }
 
         /// <summary>
-        /// Preorder folders and items that are provided as ordered evaluated includes
+        /// Get the display path for an item. The display path is what you see visually in solution explorer.
         /// </summary>
-        private void ComputeIndices()
+        /// <param name="item"></param>
+        /// <returns></returns>
+        public static string GetDisplayPath(UnconfiguredProject project, ProjectItemIdentity item)
         {
-            var duplicateFiles = OrderedItems
-                .Select(p => Path.GetFileName(p.EvaluatedInclude))
-                .GroupBy(file => file, StringComparer.OrdinalIgnoreCase)
-                .Where(g => g.Count() > 1)
-                .Select(g => g.Key).ToImmutableHashSet();
-
-            var index = 1;
-
-            // This enumerates all the folder names and file names and maps them
-            // against their assigned display index. Any file names that are duplicates
-            // are mapped in in _rootedOrderMap only; everything else is mapped in
-            // _displayOrderMap only.
-            foreach (var item in OrderedItems)
+            var linkPath = FileItemServices.GetLinkFilePath(item.Metadata);
+            
+            if (!string.IsNullOrWhiteSpace(linkPath))
             {
-                var includeParts = item.EvaluatedInclude.Split(CommonConstants.PathSeparators, StringSplitOptions.RemoveEmptyEntries);
-                foreach (var part in includeParts)
+                // This is a linked file.
+                // We use the link path because that is the rendering/display path in solution explorer.
+                return project.MakeRelative(linkPath);
+            }
+
+            return project.MakeRelative(item.EvaluatedInclude);
+        }
+
+        /// <summary>
+        /// Create an ordered map.
+        /// </summary>
+        private static Dictionary<string, int> CreateOrderedMap(UnconfiguredProject project, IReadOnlyCollection<ProjectItemIdentity> orderedItems)
+        {
+            var displayOrder = 1;
+            var orderedMap = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var item in orderedItems)
+            {
+                var displayPath = GetDisplayPath(project, item);
+                var folders = GetPathFolders(displayPath);
+
+                // We need assign the display order to folders first before the file.
+                // These folders could be physical or virtual. Virtual coming from link paths.
+                foreach (var folder in folders)
                 {
-                    var rootedPath = duplicateFiles.Contains(part) ? _project.MakeRooted(item.EvaluatedInclude) : null;
-                    if (rootedPath != null && !_rootedOrderMap.ContainsKey(rootedPath))
+                    // Folders are special. 
+                    // FIXME: Due to the lack of metadata/info from property context 
```

Comment:
What changes needs to be made to CPS to get this information?

Reply:
CPS needs to give us a better way of identifying folders by their display/rendering path.